### PR TITLE
[WIP] Attempting to fix #135

### LINKED ISSUE #135: upgrade fails with: spec.users.passwd in body must be of type string: "null"
TL;Dr I think this is because the bootstrap API doesn't have `omitempty` for `spec.users.passwd`. I can open an issue there if that's the case, but it's not clear to me yet because it seems like CAPA doesn't set `users.spec.passwd` either but it doesn't seem to be an issue there.

From the upgrade tool using CAPV w/ SSH keys for user access:
```
INFO[0000] Updating machines
INFO[0000] Running etcdctl                               args="member list -w json" pod=etcd-target-cluster02-controlplane-0
INFO[0000] etcdctl stdout: {"header":{"cluster_id":9521744471245018578,"member_id":7742574140227613697,"raft_term":2},"members":[{"ID":7742574140227613697,"name":"target-cluster02-controlplane-0","peerURLs":["https://192.168.3.98:2380"],"clientURLs":["https://192.168.3.98:2379"]}]}
INFO[0000] etcdctl stderr:
INFO[0000] Storing upgrade ID on machine                 machine=default/target-cluster02-controlplane-0 upgrade-id=test01
INFO[0000] Checking etcd health                          machine=default/target-cluster02-controlplane-0 upgrade-id=test01
INFO[0000] Running etcdctl                               args="member list -w json" pod=etcd-target-cluster02-controlplane-0
INFO[0000] etcdctl stdout: {"header":{"cluster_id":9521744471245018578,"member_id":7742574140227613697,"raft_term":2},"members":[{"ID":7742574140227613697,"name":"target-cluster02-controlplane-0","peerURLs":["https://192.168.3.98:2380"],"clientURLs":["https://192.168.3.98:2379"]}]}
INFO[0000] etcdctl stderr:
INFO[0000] Running etcdctl                               args="endpoint health --endpoints https://192.168.3.98:2379" pod=etcd-target-cluster02-controlplane-0
INFO[0000] etcdctl stdout: https://192.168.3.98:2379 is healthy: successfully committed proposal: took = 761.531µs
https://192.168.3.98:2379 is healthy: successfully committed proposal: took = 560.745µs
INFO[0000] etcdctl stderr:
INFO[0000] Updating infrastructure reference             api-version=infrastructure.cluster.x-k8s.io/v1alpha2 kind=VSphereMachine machine=default/target-cluster02-controlplane-0 name=target-cluster02-controlplane-0 upgrade-id=test01
INFO[0000] Updating bootstrap reference                  api-version=bootstrap.cluster.x-k8s.io/v1alpha2 kind=KubeadmConfig machine=default/target-cluster02-controlplane-0 name=target-cluster02-controlplane-0 upgrade-id=test01
Error: KubeadmConfig.bootstrap.cluster.x-k8s.io "target-cluster02-controlplane-0.upgrade.test01" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"bootstrap.cluster.x-k8s.io/v1alpha2", "kind":"KubeadmConfig", "metadata":map[string]interface {}{"annotations":map[string]interface {}{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"bootstrap.cluster.x-k8s.io/v1alpha2\",\"kind\":\"KubeadmConfig\",\"metadata\":{\"annotations\":{},\"name\":\"target-cluster02-controlplane-0\",\"namespace\":\"default\"},\"spec\":{\"clusterConfiguration\":{\"apiServer\":{\"extraArgs\":{\"cloud-provider\":\"external\"}},\"controllerManager\":{\"extraArgs\":{\"cloud-provider\":\"external\"}},\"imageRepository\":\"k8s.gcr.io\"},\"initConfiguration\":{\"nodeRegistration\":{\"criSocket\":\"/var/run/containerd/containerd.sock\",\"kubeletExtraArgs\":{\"cloud-provider\":\"external\"},\"name\":\"{{ ds.meta_data.hostname }}\"}},\"preKubeadmCommands\":[\"hostname \\\"{{ 
....
....
....
validation failure list:
spec.users.passwd in body must be of type string: "null"
```

Adding an empty string to `spec.users.passwd` field of the control plane KubeadmConfig fixed it. I thought it was weird that `spec.users.passwd` [doesn't have the omitempty json tag](https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/api/v1alpha2/kubeadmbootstrapconfig_types.go#L171). I think that's the correct fix but I don't have history/context on that.

### PR DESCRIPTION
**Pull request by exagil**
I was trying to upgrade the cluster while I encountered issue #135 
This required me to mark the `passwd` field optional, which was already done in the latest versions. Hence, I've attempted to migrate to the latest version of dependencies.

Primarily, since `sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm` was deprecated and merged with `sigs.k8s.io/cluster-api`, this PR upgrades the `sigs.k8s.io/cluster-api` version, removes the dependency of `sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm` and gets the build to pass.

In the attempt to upgrade the version, one of the error messages required me to upgrade the cluster version too, and I'm not sure of the side effects of doing it. Please suggest how I should go about it.

### CODE REVIEW

**Comment by exagil** [comment]
@ncdc @randomvariable

File: pkg/upgrade/control_plane.go [hunk 1]
```diff
@@ -29,9 +29,10 @@ import (
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
-	bootstrapv1 "sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm/api/v1alpha2"
-	kubeadmv1beta1 "sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm/kubeadm/v1beta1"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	clusterv1a3 "sigs.k8s.io/cluster-api/api/v1alpha3"
```
Comment (ncdc) [inline pkg/upgrade/control_plane.go]: This tool is only for v1alpha2 and it is not applicable for v1alpha3 clusters (upgrade functionality is being built in to v1alpha3). The path forward to fix the issue is to make a pull request to the deprecated cluster-api-bootstrap-provider-kubeadm repository, adjusting the `omitempty` tags in the kubeadm types. Once we merge that pull request and tag a new release, you can update this one to pull in the the new tag.

**Comment by ncdc** [comment]
We have filed https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/issues/291 to fix the omitempty tags in CABPK. Once that's done, and we have a new release (https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/issues/292), please update this pull request to use CABPK v0.1.6. You should not need to make any other code changes. Thanks!

File: pkg/upgrade/control_plane.go (hunk 1 above)
Comment (exagil) [reply pkg/upgrade/control_plane.go]: Sure. Will do that.

Comment (ncdc) [reply pkg/upgrade/control_plane.go]: We have a new CABPK release, so you should be able to proceed now. Thanks!

**Comment by ncdc** [comment]
Replaced by #169

**Pull request closed without merging**